proposals/NNNN-spirv-types-in-clang.md

```diff
@@ -0,0 +1,113 @@
+
+We are implementing `SpirvType` and `SpirvOpaqueType` defined in
+[Inline SPIR-V (HLSL proposal 0011)](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0011-inline-spirv.md#types).
+In order to support this feature, we need a way to represent inline SPIR-V types
+in the Clang type system. To facilitate this, I am proposing
+`HLSLInlineSpirvType` as a new `Type` subclass, and `__hlsl_spirv_type` as a new
+builtin template to create such a type.
+
+## Motivation
+
+In DXC, we represented `SpirvType` as a template for a struct. This was simple
+to implement, since in DXC we lower the AST directly to SPIR-V without going
+through LLVM codegen, and could add special cases handling structs that were
+`SpirvType` specializations while lowering types. However, this approach does
+not interact well with Sema and the rest of Clang, resulting in at least several
+bugs in handling template arguments and other areas. In order to take the same
+approach in Clang, we would need to add special case handling while lowering
+struct types, as well as for calculating size and alignment and anywhere else
+structs are interacted with in Clang. Struct types are already processed
+separately from all other types while lowering, and adding special cases for
+`SpirvType` would complicate their handling even more.
+
+Finally, a struct is not a good semantic representation of a `SpirvType`.
+`SpirvType` is fundamentally a new kind of type for Clang that is different from
+all existing types—it represents a low-level target type specified by the user.
+
+## Proposed solution
+
+### Type Representation
+
+Describe your solution to the problem. Provide examples and describe how they
+work. Show how your solution is better than current workarounds: is it cleaner,
+safer, or more efficient?
+
+`HLSLInlineSpirvType` is a subclass of `clang::Type`, which has the properties
+
+| Property    | Optional | Type                   |
+| ----------- | -------- | ---------------------- |
+| `Operand`   | no       | `uint32_t`             |
+| `Size`      | yes      | `uint32_t`             |
+| `Alignment` | yes      | `uint32_t`             |
+| `Operands`  | no       | array of`SpirvOperand` |
+
+A value of type `HLSLInlineSpirvType` is always canonical.
+
+#### Operands
+
+Each operand in `Operands` is a value of the nested class
+`HLSLInlineSpirvType::SpirvOperand`, which can be one of the following kinds:
+
+| Kind         | Description                                                                        | Has Result Type | Has Integral Value |
+| ------------ | ---------------------------------------------------------------------------------- | --------------- | ------------------ |
+| `ConstantId` | Represents a value to be passed in as the ID of a SPIR-V `OpConstant` instruction. | yes             | yes                |
+| `Literal`    | Represents a value to be passed in as an immediate literal.                        | no              | yes                |
```

Comment:
Would this support "string" literals?
Ex using this to represent: `OpTypeImage [...] 2D [...] Unknown` ?
Or can you use `OpUnknown` to encode arbitrary operands like `DIM` ?

Reply:
Those are not string operands in the instruction. They are enums and are represented using the Literal operand. We don't want to get into converting the string in the comipler because a new enum may be added that the compiler does not know about. It would also be a break from DXC.